dashboard/app/helpers/levels_helper.rb

```diff
@@ -170,30 +171,32 @@ def app_options
       view_options(authored_hint_view_requests_url: authored_hint_view_requests_path(format: :json))
     end
 
-    if @level.is_a? Blockly
-      @app_options = blockly_options
-    elsif @level.is_a?(DSLDefined) || @level.is_a?(FreeResponse)
-      @app_options = question_options
-    elsif @level.is_a? Widget
-      @app_options = widget_options
-    elsif @level.unplugged?
-      @app_options = unplugged_options
+    if level.is_a? Blockly
+      @app_options[level.id] = blockly_options level
+    elsif level.is_a?(DSLDefined) || level.is_a?(FreeResponse)
+      @app_options[level.id] = question_options level
+    elsif level.is_a? Widget
+      @app_options[level.id] = widget_options level
+    elsif level.unplugged?
+      @app_options[level.id] = unplugged_options
     else
       # currently, all levels are Blockly or DSLDefined except for Unplugged
-      @app_options = view_options.camelize_keys
+      @app_options[level.id] = view_options.camelize_keys
     end
+    reset_view_options
```

Comment:
one incremental change to make this saner would be to change `view_options()` (no arguments) to `freeze_view_options` so that it's obvious when it's getting frozen vs. being called in a non-freezing manner